File: SteamAPIWrap/GameServer.cpp
#include "Precompiled.hpp"
#include "GameServer.hpp"

#include "OverlayDialog.hpp"
#include "OverlayDialogToUser.hpp"

#include "MemoryHelpers.h"


using namespace SteamAPIWrap;

namespace SteamAPIWrap
{
	template <> CGameServer* CSingleton<CGameServer>::instance = nullptr;

	CGameServer::CGameServer() 
		: nativeSteamServersConnectedCallback(this, &CGameServer::OnSteamServersConnected)
		, nativeSteamServerConnectFailureCallback(this, &CGameServer::OnSteamServerConnectFailure)
		, nativeSteamServersDisconnectedCallback(this, &CGameServer::OnSteamServersDisconnected)
		, nativeValidateAuthTicketResponseCallback(this, &CGameServer::OnValidateAuthTicketResponse)
		, nativeGSClientApproveCallback(this, &CGameServer::OnGSClientApprove)
		, nativeGSClientDenyCallback(this, &CGameServer::OnGSClientDeny)
		, nativeGSClientKickCallback(this, &CGameServer::OnGSClientKick)
		, nativeGSClientAchievementStatusCallback(this, &CGameServer::OnGSClientAchievementStatus)
		, nativeGSPolicyResponseCallback(this, &CGameServer::OnGSPolicyResponse)
		, nativeGSGameplayStatsCallback(this, &CGameServer::OnGSGameplayStats)
		, nativeGSClientGroupStatusCallback(this, &CGameServer::OnGSClientGroupStatus)
		, gameServer(nullptr)
	{
		
	}

	bool CGameServer::InitGameServer( u32 ip, u16 gamePort, u16 queryPort, u32 flags, 
		AppID gameAppId, PConstantString versionString )
	{
		return gameServer->InitGameServer(ip,gamePort,queryPort,flags,gameAppId,versionString);
	}

	void CGameServer::SetProduct( PConstantString product )
	{
		gameServer->SetProduct(product);
	}

	void CGameServer::SetGameDescription( PConstantString gameDescription )
	{
		gameServer->SetGameDescription(gameDescription);
	}

	void CGameServer::SetModDir( PConstantString modDir )
	{
		gameServer->SetModDir(modDir);
	}

	void CGameServer::SetDedicatedServer( bool dedicated )
	{
		gameServer->SetDedicatedServer(dedicated);
	}

	void CGameServer::LogOn(PConstantString  accountName, PConstantString password )
	{
		gameServer->LogOn(accountName,password);
	}

	void CGameServer::LogOnAnonymous()
	{
		gameServer->LogOnAnonymous();
	}

	void CGameServer::LogOff()
	{
		gameServer->LogOff();
	}

	bool CGameServer::BLoggedOn()
	{
		return gameServer->BLoggedOn();
	}


	bool CGameServer::BSecure()
	{
		return gameServer->BSecure();
	}

	bool CGameServer::WasRestartRequested()
	{
		return gameServer->WasRestartRequested();
	}

	SteamID CGameServer::GetSteamID()
	{
		return gameServer->GetSteamID().ConvertToUint64();
	}

	void CGameServer::SetMaxPlayerCount( int cPlayersMax )
	{
		gameServer->SetMaxPlayerCount(cPlayersMax);
	}

	void CGameServer::SetBotPlayerCount( int cBotplayers )
	{
		gameServer->SetBotPlayerCount(cBotplayers);
	}

	void CGameServer::SetServerName(PConstantString serverName )
	{
		gameServer->SetServerName(serverName);
	}

	void CGameServer::SetMapName(PConstantString  mapName )
	{
		gameServer->SetMapName(mapName);
	}

	void CGameServer::SetPasswordProtected( bool passwordProtected )
	{
		gameServer->SetPasswordProtected(passwordProtected);
	}

	void CGameServer::SetSpectatorPort( u16 spectatorPort )
	{
		gameServer->SetSpectatorPort(spectatorPort);
	}

	void CGameServer::SetSpectatorServerName(PConstantString  spectatorServerName )
	{
		gameServer->SetSpectatorServerName(spectatorServerName);
	}

	void CGameServer::ClearAllKeyValues()
	{
		gameServer->ClearAllKeyValues();
	}

	void CGameServer::SetKeyValue( PConstantString key, PConstantString value )
	{
		gameServer->SetKeyValue(key, value);
	}

	void CGameServer::SetGameTags( PConstantString gameTags )
	{
		gameServer->SetGameTags(gameTags);
	}

	void CGameServer::SetGameData( PConstantString gameData )
	{
		gameServer->SetGameData(gameData);
	}

	void CGameServer::SetRegion( PConstantString region )
	{
		gameServer->SetRegion(region);
	}

	bool CGameServer::SendUserConnectAndAuthenticate( u32 ipClient, PConstantDataPointer authBlob, u32 authBlobSize, SteamID *steamIDUser )
	{
		CSteamID idUser;
		bool results = gameServer->SendUserConnectAndAuthenticate(ipClient, authBlob, authBlobSize, &idUser);
		*steamIDUser = idUser.ConvertToUint64();
		return results;
	}

	SteamID CGameServer::CreateUnauthenticatedUserConnection()
	{
		return gameServer->CreateUnauthenticatedUserConnection().ConvertToUint64();
	}

	void CGameServer::SendUserDisconnect( SteamID steamIDUser )
	{
		gameServer->SendUserDisconnect(steamIDUser);
	}

	bool CGameServer::BUpdateUserData( SteamID steamIDUser, PConstantString  playerName, u32 core )
	{
		return gameServer->BUpdateUserData(steamIDUser, playerName, core);
	}

	HAuthTicket CGameServer::GetAuthSessionTicket( PDataPointer ticket, int maxTicket, uint32 *ticketSize )
	{
		return gameServer->GetAuthSessionTicket(ticket, maxTicket, ticketSize);
	}

	Enum CGameServer::BeginAuthSession( PConstantDataPointer authTicket, s32 authTicketSize, SteamID steamID )
	{
		return gameServer->BeginAuthSession(authTicket, authTicketSize, steamID);
	}

	void CGameServer::EndAuthSession(SteamID steamID )
	{
		gameServer->EndAuthSession(steamID);
	}

	void CGameServer::CancelAuthTicket( u32 authTicket )
	{
		gameServer->CancelAuthTicket(authTicket);
	}

	Enum CGameServer::UserHasLicenseForApp( SteamID steamID, u32 appID )
	{
		return gameServer->UserHasLicenseForApp(steamID, appID);
	}

	bool CGameServer::RequestUserGroupStatus( SteamID steamIDUser, SteamID steamIDGroup )
	{
		return gameServer->RequestUserGroupStatus(steamIDUser, steamIDGroup);
	}

	void CGameServer::GetGameplayStats()
	{
		gameServer->GetGameplayStats();
	}

	void CGameServer::GetServerReputation()
	{
		resultGSReputation.Set(gameServer->GetServerReputation(),this, &CGameServer::OnGSReputation);
	}

	uint32 CGameServer::GetPublicIP()
	{
		return gameServer->GetPublicIP();
	}

	bool CGameServer::HandleIncomingPacket( PConstantDataPointer data, int dataSize, u32 ip, u16 srcPort )
	{
		return gameServer->HandleIncomingPacket(data, dataSize, ip, srcPort);
	}

	int CGameServer::GetNextOutgoingPacket( PDataPointer out, int maxOut, u32 *netAdr, u16 *port )
	{
		return gameServer->GetNextOutgoingPacket(out, maxOut, netAdr, port);
	}

	void CGameServer::EnableHeartbeats( bool active )
	{
		gameServer->EnableHeartbeats(active);
	}

	void CGameServer::SetHeartbeatInterval( int heartbeatInterval )
	{
		gameServer->SetHeartbeatInterval(heartbeatInterval);
	}

	void CGameServer::ForceHeartbeat()
	{
		gameServer->ForceHeartbeat();
	}

	void CGameServer::AssociateWithClan( SteamID steamIDClan )
	{
		resultAssociateWithClanResult.Set(gameServer->AssociateWithClan(SteamID(steamIDClan)), 
			this, &CGameServer::OnAssociateWithClanResult);
	}

	void CGameServer::ComputeNewPlayerCompatibility( SteamID steamIDNewPlayer )
	{
		resultComputeNewPlayerCompatibilityResult.Set(gameServer->ComputeNewPlayerCompatibility(steamIDNewPlayer),
			this, &CGameServer::OnComputeNewPlayerCompatibilityResult);
	}
}

ManagedSteam_API bool GameServer_InitGameServer( u32 ip, u16 gamePort, u16 queryPort, u32 flags, 
	AppID gameAppId, PConstantString versionString )
{
	return CGameServer::Instance().InitGameServer(ip,gamePort,queryPort,flags,gameAppId,versionString);

}

ManagedSteam_API void GameServer_SetProduct( PConstantString product )
{
	CGameServer::Instance().SetProduct(product);
}

ManagedSteam_API void GameServer_SetGameDescription(PConstantString  gameDescription )
{
	CGameServer::Instance().SetGameDescription(gameDescription);
}

ManagedSteam_API void GameServer_SetModDir( PConstantString modDir )
{
	CGameServer::Instance().SetModDir(modDir);
}

ManagedSteam_API void GameServer_SetDedicatedServer( bool dedicated )
{
	CGameServer::Instance().SetDedicatedServer(dedicated);
}

ManagedSteam_API void GameServer_LogOn(PConstantString accountName, PConstantString password )
{
	CGameServer::Instance().LogOn(accountName,password);
}

ManagedSteam_API void GameServer_LogOnAnonymous()
{
	CGameServer::Instance().LogOnAnonymous();
}

ManagedSteam_API void GameServer_LogOff()
{
	CGameServer::Instance().LogOff();
}

ManagedSteam_API bool GameServer_LoggedOn()
{
	return CGameServer::Instance().BLoggedOn();
}

ManagedSteam_API bool GameServer_Secure()
{
	return CGameServer::Instance().BSecure();
}

ManagedSteam_API SteamID GameServer_GetSteamID()
{
	return CGameServer::Instance().GetSteamID();
}

ManagedSteam_API bool GameServer_WasRestartRequested()
{
	return CGameServer::Instance().WasRestartRequested();
}

ManagedSteam_API void GameServer_SetMaxPlayerCount( int cPlayersMax )
{
	CGameServer::Instance().SetMaxPlayerCount(cPlayersMax);
}

ManagedSteam_API void GameServer_SetBotPlayerCount( int cBotplayers )
{
	CGameServer::Instance().SetBotPlayerCount(cBotplayers);
}

ManagedSteam_API void GameServer_SetServerName(PConstantString  serverName )
{
	CGameServer::Instance().SetServerName(serverName);
}

ManagedSteam_API void GameServer_SetMapName(PConstantString  mapName )
{
	CGameServer::Instance().SetMapName(mapName);
}

ManagedSteam_API void GameServer_SetPasswordProtected( bool passwordProtected )
{
	CGameServer::Instance().SetPasswordProtected(passwordProtected);
}

ManagedSteam_API void GameServer_SetSpectatorPort( u16 spectatorPort )
{
	CGameServer::Instance().SetSpectatorPort(spectatorPort);
}

ManagedSteam_API void GameServer_SetSpectatorServerName(PConstantString spectatorServerName )
{
	CGameServer::Instance().SetSpectatorServerName(spectatorServerName);
}

ManagedSteam_API void GameServer_ClearAllKeyValues()
{
	CGameServer::Instance().ClearAllKeyValues();
}

ManagedSteam_API void GameServer_SetKeyValue( PConstantString key, PConstantString value )
{
	CGameServer::Instance().SetKeyValue(key, value);
}

ManagedSteam_API void GameServer_SetGameTags( PConstantString gameTags )
{
	CGameServer::Instance().SetGameTags(gameTags);
}

ManagedSteam_API void GameServer_SetGameData( PConstantString gameData )
{
	CGameServer::Instance().SetGameData(gameData);
}

ManagedSteam_API void GameServer_SetRegion( PConstantString region )
{
	CGameServer::Instance().SetRegion(region);
}

ManagedSteam_API bool GameServer_SendUserConnectAndAuthenticate( u32 ipClient, PConstantDataPointer authBlob, u32 authBlobSize, SteamID *steamIDUser )
{
	return CGameServer::Instance().SendUserConnectAndAuthenticate(ipClient,authBlob,authBlobSize,steamIDUser);
}

ManagedSteam_API SteamID GameServer_CreateUnauthenticatedUserConnection()
{
	return CGameServer::Instance().CreateUnauthenticatedUserConnection();
}

ManagedSteam_API void GameServer_SendUserDisconnect( SteamID steamIDUser )
{
	CGameServer::Instance().SendUserDisconnect(steamIDUser);
}

ManagedSteam_API bool GameServer_BUpdateUserData( SteamID steamIDUser, PConstantString  playerName, u32 core )
{
	return CGameServer::Instance().BUpdateUserData(steamIDUser,playerName,core);
}

ManagedSteam_API HAuthTicket GameServer_GetAuthSessionTicket( PDataPointer ticket, int maxTicket, uint32 *ticketSize )
{
	return CGameServer::Instance().GetAuthSessionTicket(ticket,maxTicket,ticketSize);
}

ManagedSteam_API Enum GameServer_BeginAuthSession( PConstantDataPointer authTicket, s32 authTicketSize, SteamID steamID )
{
	return (int)CGameServer::Instance().BeginAuthSession(authTicket, authTicketSize, steamID);
}

ManagedSteam_API void GameServer_EndAuthSession(SteamID steamID )
{
	CGameServer::Instance().EndAuthSession(steamID);
}

ManagedSteam_API void GameServer_CancelAuthTicket( u32 authTicket )
{
	CGameServer::Instance().CancelAuthTicket(authTicket);
}

ManagedSteam_API Enum GameServer_UserHasLicenseForApp( SteamID steamID, u32 appID )
{
	return (int)CGameServer::Instance().UserHasLicenseForApp(steamID,appID);
}

ManagedSteam_API bool GameServer_RequestUserGroupStatus( SteamID steamIDUser, SteamID steamIDGroup )
{
	return CGameServer::Instance().RequestUserGroupStatus(steamIDUser,steamIDGroup);
}

ManagedSteam_API void GameServer_GetGameplayStats()
{
	CGameServer::Instance().GetGameplayStats();
}

ManagedSteam_API void GameServer_GetServerReputation()
{
	CGameServer::Instance().GetServerReputation();
}

ManagedSteam_API uint32 GameServer_GetPublicIP()
{
	return CGameServer::Instance().GetPublicIP();
}

ManagedSteam_API bool GameServer_HandleIncomingPacket( PConstantDataPointer data, int dataSize, u32 ip, u16 srcPort )
{
	return CGameServer::Instance().HandleIncomingPacket(data,dataSize,ip,srcPort);
}

ManagedSteam_API int GameServer_GetNextOutgoingPacket( PDataPointer out, int maxOut, u32 *netAdr, u16 *port )
{
	return CGameServer::Instance().GetNextOutgoingPacket(out,maxOut,netAdr,port);
}

ManagedSteam_API void GameServer_EnableHeartbeats( bool active )
{
	CGameServer::Instance().EnableHeartbeats(active);
}

ManagedSteam_API void GameServer_SetHeartbeatInterval( int heartbeatInterval )
{
	CGameServer::Instance().SetHeartbeatInterval(heartbeatInterval);
}

ManagedSteam_API void GameServer_ForceHeartbeat()
{
	CGameServer::Instance().ForceHeartbeat();
}

ManagedSteam_API void GameServer_AssociateWithClan( SteamID steamIDClan )
{
	CGameServer::Instance().AssociateWithClan(steamIDClan);
}

ManagedSteam_API void GameServer_ComputeNewPlayerCompatibility( SteamID steamIDNewPlayer )
{
	CGameServer::Instance().ComputeNewPlayerCompatibility(steamIDNewPlayer);
}

File: ManagedSteam/Implementations/GameServer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;

using ManagedSteam.CallbackStructures;
using ManagedSteam.SteamTypes;
using ManagedSteam.Utility;

namespace ManagedSteam.Implementations
{
    class GameServer : SteamServiceGameServer, IGameServer
    {
        private List<SteamServersConnected> steamServersConnected;
        private List<SteamServerConnectFailure> steamServerConnectFailure;
        private List<SteamServersDisconnected> steamServersDisconnected;
        private List<ValidateAuthTicketResponse> validateAuthTicketResponse;

        private List<GSClientApprove> gsClientApprove;
        private List<GSClientDeny> gsClientDeny;
        private List<GSClientKick> gsClientKick;
        private List<GSClientAchievementStatus> gsClientAchievementStatus;
        private List<GSPolicyResponse> gsPolicyResponse;
        private List<GSGameplayStats> gsGameplayStats;
        private List<GSClientGroupStatus> gsClientGroupStatus;

        private List<Result<GSReputation>> gsReputation;
        private List<Result<AssociateWithClanResult>> associateWithClanResult;
        private List<Result<ComputeNewPlayerCompatibilityResult>> computeNewPlayerCompatibilityResult;

        public GameServer()
        {
            steamServersConnected = new List<SteamServersConnected>();
            steamServerConnectFailure = new List<SteamServerConnectFailure>();
            steamServersDisconnected = new List<SteamServersDisconnected>();
            validateAuthTicketResponse = new List<ValidateAuthTicketResponse>();

            gsClientApprove = new List<GSClientApprove>();
            gsClientDeny = new List<GSClientDeny>();
            gsClientKick = new List<GSClientKick>();
            gsClientAchievementStatus = new List<GSClientAchievementStatus>();
            gsPolicyResponse = new List<GSPolicyResponse>();
            gsGameplayStats = new List<GSGameplayStats>();
            gsClientGroupStatus = new List<GSClientGroupStatus>();

            gsReputation = new List<Result<GSReputation>>();
            associateWithClanResult = new List<Result<AssociateWithClanResult>>();
            computeNewPlayerCompatibilityResult = new List<Result<ComputeNewPlayerCompatibilityResult>>();


            Callbacks[CallbackID.SteamServersConnected] = (data, size) => steamServersConnected.Add(CallbackStructures.SteamServersConnected.Create(data, size));
            Callbacks[CallbackID.SteamServerConnectFailure] = (data, size) => steamServerConnectFailure.Add(CallbackStructures.SteamServerConnectFailure.Create(data, size));
            Callbacks[CallbackID.SteamServersDisconnected] = (data, size) => steamServersDisconnected.Add(CallbackStructures.SteamServersDisconnected.Create(data, size));
            Callbacks[CallbackID.ValidateAuthTicketResponse] = (data, size) => validateAuthTicketResponse.Add(CallbackStructures.ValidateAuthTicketResponse.Create(data, size));

            Callbacks[CallbackID.GSClientApprove] = (data, size) => gsClientApprove.Add(CallbackStructures.GSClientApprove.Create(data, size));
            Callbacks[CallbackID.GSClientDeny] = (data, size) => gsClientDeny.Add(CallbackStructures.GSClientDeny.Create(data, size));
            Callbacks[CallbackID.GSClientKick] = (data, size) => gsClientKick.Add(CallbackStructures.GSClientKick.Create(data, size));
            Callbacks[CallbackID.GSClientAchievementStatus] = (data, size) => gsClientAchievementStatus.Add(CallbackStructures.GSClientAchievementStatus.Create(data, size));
            Callbacks[CallbackID.GSPolicyResponse] = (data, size) => gsPolicyResponse.Add(CallbackStructures.GSPolicyResponse.Create(data, size));
            Callbacks[CallbackID.GSGameplayStats] = (data, size) => gsGameplayStats.Add(CallbackStructures.GSGameplayStats.Create(data, size));
            Callbacks[CallbackID.GSClientGroupStatus] = (data, size) => gsClientGroupStatus.Add(CallbackStructures.GSClientGroupStatus.Create(data, size));

            Results[ResultID.GSReputation] = (data, size, flag) => gsReputation.Add(new Result<GSReputation>(CallbackStructures.GSReputation.Create(data, size), flag));
            Results[ResultID.AssociateWithClanResult] = (data, size, flag) => associateWithClanResult.Add(new Result<AssociateWithClanResult>(CallbackStructures.AssociateWithClanResult.Create(data, size), flag));
            Results[ResultID.ComputeNewPlayerCompatibilityResult] = (data, size, flag) => computeNewPlayerCompatibilityResult.Add(new Result<ComputeNewPlayerCompatibilityResult>(CallbackStructures.ComputeNewPlayerCompatibilityResult.Create(data, size), flag));
        }

        public event CallbackEvent<SteamServersConnected> SteamServersConnected;
        public event CallbackEvent<SteamServerConnectFailure> SteamServerConnectFailure;
        public event CallbackEvent<SteamServersDisconnected> SteamServersDisconnected;
        public event CallbackEvent<ValidateAuthTicketResponse> ValidateAuthTicketResponse;

        public event CallbackEvent<GSClientApprove> GSClientApprove;
        public event CallbackEvent<GSClientDeny> GSClientDeny;
        public event CallbackEvent<GSClientKick> GSClientKick;
        public event CallbackEvent<GSClientAchievementStatus> GSClientAchievementStatus;
        public event CallbackEvent<GSPolicyResponse> GSPolicyResponse;
        public event CallbackEvent<GSGameplayStats> GSGameplayStats;
        public event CallbackEvent<GSClientGroupStatus> GSClientGroupStatus;

        public event ResultEvent<GSReputation> GSReputation;
        public event ResultEvent<AssociateWithClanResult> AssociateWithClanResult;
        public event ResultEvent<ComputeNewPlayerCompatibilityResult> ComputeNewPlayerCompatibilityResult;


        internal override void CheckIfUsableInternal()
        {

        }

        internal override void ReleaseManagedResources()
        {
            steamServersConnected = null;
            SteamServersConnected = null;
            steamServerConnectFailure = null;
            SteamServerConnectFailure = null;
            steamServersDisconnected = null;
            SteamServersDisconnected = null;
            validateAuthTicketResponse = null;
            ValidateAuthTicketResponse = null;

            gsClientApprove = null;
            GSClientApprove = null;
            gsClientDeny = null;
            GSClientDeny = null;
            gsClientKick = null;
            GSClientKick = null;
            gsClientAchievementStatus = null;
            GSClientAchievementStatus = null;
            gsPolicyResponse = null;
            GSPolicyResponse = null;
            gsGameplayStats = null;
            GSGameplayStats = null;
            gsClientGroupStatus = null;
            GSClientGroupStatus = null;

            gsReputation = null;
            GSReputation = null;
            associateWithClanResult = null;
            AssociateWithClanResult = null;
            computeNewPlayerCompatibilityResult = null;
            ComputeNewPlayerCompatibilityResult = null;
        }

        internal override void InvokeEvents()
        {
            InvokeEvents(steamServersConnected, SteamServersConnected);
            InvokeEvents(steamServerConnectFailure, SteamServerConnectFailure);
            InvokeEvents(steamServersDisconnected, SteamServersDisconnected);
            InvokeEvents(validateAuthTicketResponse, ValidateAuthTicketResponse);

            InvokeEvents(gsClientApprove, GSClientApprove);
            InvokeEvents(gsClientDeny, GSClientDeny);
            InvokeEvents(gsClientKick, GSClientKick);
            InvokeEvents(gsClientAchievementStatus, GSClientAchievementStatus);
            InvokeEvents(gsPolicyResponse, GSPolicyResponse);
            InvokeEvents(gsGameplayStats, GSGameplayStats);
            InvokeEvents(gsClientGroupStatus, GSClientGroupStatus);

            InvokeEvents(gsReputation, GSReputation);
            InvokeEvents(associateWithClanResult, AssociateWithClanResult);
            InvokeEvents(computeNewPlayerCompatibilityResult, ComputeNewPlayerCompatibilityResult);
        }

        public bool InitGameServer(uint ip, ushort gamePort, ushort queryPort, uint flags, AppID gameAppId, string versionString)
        {
            //CheckIfUsable();
            return NativeMethods.GameServer_InitGameServer(ip, gamePort, queryPort, flags, gameAppId.AsUInt32, versionString);
        }

        public void SetProduct(string product)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetProduct(product);
        }

        public void SetGameDescription(string gameDescription)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetGameDescription(gameDescription);
        }

        public void SetModDir(string modDir)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetModDir(modDir);
        }

        public void SetDedicatedServer(bool dedicated)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetDedicatedServer(dedicated);
        }

        public void LogOn(string accountName, string password)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_LogOn(accountName, password);
        }

        public void LogOnAnonymous()
        {
            //CheckIfUsable();
            NativeMethods.GameServer_LogOnAnonymous();
        }

        public void LogOff()
        {
            //CheckIfUsable();
            NativeMethods.GameServer_LogOff();
        }

        public bool LoggedOn()
        {
            //CheckIfUsable();
            return NativeMethods.GameServer_LoggedOn();
        }

        public bool Secure()
        {
            //CheckIfUsable();
            return NativeMethods.GameServer_Secure();
        }

        public SteamID GetSteamID()
        {
            //CheckIfUsable();
            return new SteamID(NativeMethods.GameServer_GetSteamID());
        }

        public bool WasRestartRequested()
        {
            //CheckIfUsable();
            return NativeMethods.GameServer_WasRestartRequested();
        }

        public void SetMaxPlayerCount(int playersMax)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetMaxPlayerCount(playersMax);
        }

        public void SetBotPlayerCount(int botPlayers)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetBotPlayerCount(botPlayers);
        }

        public void SetServerName(string serverName)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetServerName(serverName);
        }

        public void SetMapName(string mapName)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetMapName(mapName);
        }

        public void SetPasswordProtected(bool passwordProtected)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetPasswordProtected(passwordProtected);
        }

        public void SetSpectatorPort(ushort spectatorPort)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetSpectatorPort(spectatorPort);
        }

        public void SetSpectatorServerName(string spectatorServerName)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetSpectatorServerName(spectatorServerName);
        }

        public void ClearAllKeyValues()
        {
            //CheckIfUsable();
            NativeMethods.GameServer_ClearAllKeyValues();
        }

        public void SetKeyValue(string key, string value)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetKeyValue(key, value);
        }

        public void SetGameTags(string gameTags)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetGameTags(gameTags);
        }

        public void SetGameData(string gameData)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetGameData(gameData);
        }

        public void SetRegion(string region)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetRegion(region);
        }

        public bool SendUserConnectAndAuthenticate(uint ipClient, byte[] authenticationBlob, out SteamID steamIDUser)
        {
            //CheckIfUsable();

            using (NativeBuffer blobBuffer = new NativeBuffer(authenticationBlob))
            {
                blobBuffer.WriteToUnmanagedMemory();
                ulong rawCreator = 0;
                bool result = NativeMethods.GameServer_SendUserConnectAndAuthenticate(ipClient,
                    blobBuffer.UnmanagedMemory, (uint)blobBuffer.UnmanagedMemory, ref rawCreator);

                steamIDUser = new SteamID(rawCreator);
                return result;

            }

        }

        public GameServerSendUserConnectAndAuthenticateResult SendUserConnectAndAuthenticate(uint ipClient, byte[] authenticationBlob)
        {
            GameServerSendUserConnectAndAuthenticateResult result = new GameServerSendUserConnectAndAuthenticateResult();

            result.Result = SendUserConnectAndAuthenticate(ipClient, authenticationBlob, out result.SteamIDUser);

            return result;
        }


        public SteamID CreateUnauthenticatedUserConnection()
        {
            //CheckIfUsable();
            return new SteamID(NativeMethods.GameServer_CreateUnauthenticatedUserConnection());
        }

        public void SendUserDisconnect(SteamID steamIDUser)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SendUserDisconnect(steamIDUser.AsUInt64);
        }

        public bool UpdateUserData(SteamID steamIDUser, string playerName, uint score)
        {
            //CheckIfUsable();
            return NativeMethods.GameServer_UpdateUserData(steamIDUser.AsUInt64, playerName, score);
        }

        public AuthTicketHandle GetAuthSessionTicket(System.IntPtr ticket, int maxTicket, out uint ticketLength)
        {
            //CheckIfUsable();
            ticketLength = 0;
            return new AuthTicketHandle(NativeMethods.GameServer_GetAuthSessionTicket(ticket, maxTicket, ref ticketLength));
        }

        public GameServerGetAuthSessionTicketResult GetAuthSessionTicket(System.IntPtr ticket, int maxTicket)
        {
            GameServerGetAuthSessionTicketResult result = new GameServerGetAuthSessionTicketResult();

            result.Result = GetAuthSessionTicket(ticket, maxTicket, out result.TicketSize);

            return result;
        }

        public BeginAuthSessionResult BeginAuthSession(System.IntPtr authTicket, int cbAuthTicket, SteamID steamID)
        {
            //CheckIfUsable();
            return (BeginAuthSessionResult)NativeMethods.GameServer_BeginAuthSession(authTicket, cbAuthTicket, steamID.AsUInt64);
        }

        public void EndAuthSession(SteamID steamID)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_EndAuthSession(steamID.AsUInt64);
        }

        public void CancelAuthTicket(AuthTicketHandle authTicket)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_CancelAuthTicket(authTicket.AsUInt32);
        }

        public UserHasLicenseForAppResult UserHasLicenseForApp(SteamID steamID, AppID appID)
        {
            //CheckIfUsable();
            int tempreturn = NativeMethods.GameServer_UserHasLicenseForApp(steamID.AsUInt64, appID.AsUInt32);
            return (UserHasLicenseForAppResult)tempreturn;
        }

        public bool RequestUserGroupStatus(SteamID steamIDUser, SteamID steamIDGroup)
        {
            //CheckIfUsable();
            return NativeMethods.GameServer_RequestUserGroupStatus(steamIDUser.AsUInt64, steamIDGroup.AsUInt64);
        }

        [Obsolete("As of Steamworks SDK 1.28 this function has been deprecated. It currently does not return anything, and will be removed in the future.")]
        public void GetGameplayStats()
        {
            //CheckIfUsable();
            NativeMethods.GameServer_GetGameplayStats();
        }

        [Obsolete("As of Steamworks SDK 1.28 this function has been deprecated. It currently does not return anything, and will be removed in the future.")]
        public void GetServerReputation()
        {
            //CheckIfUsable();
            NativeMethods.GameServer_GetServerReputation();
        }

        public uint GetPublicIP()
        {
            //CheckIfUsable();
            return NativeMethods.GameServer_GetPublicIP();
        }

        public bool HandleIncomingPacket(byte[] data, uint ip, ushort port)
        {
            //CheckIfUsable();
            using (NativeBuffer packetBuffer = new NativeBuffer(data))
            {
                packetBuffer.WriteToUnmanagedMemory();
                return NativeMethods.GameServer_HandleIncomingPacket(packetBuffer.UnmanagedMemory,
                    packetBuffer.UnmanagedSize, ip, port);
            }
        }

        public int GetNextOutgoingPacket(byte[] data, out uint netAdr, out ushort port)
        {
            //CheckIfUsable();
            netAdr = 0;
            port = 0;
            using (NativeBuffer packetBuffer = new NativeBuffer(data))
            {
                int result = NativeMethods.GameServer_GetNextOutgoingPacket(packetBuffer.UnmanagedMemory, packetBuffer.UnmanagedSize, ref netAdr, ref port);
                packetBuffer.ReadFromUnmanagedMemory(result);
                return result;
            }
        }

        public GameServerGetNextOutgoingPacketResult GetNextOutgoingPacket(byte[] data)
        {
            GameServerGetNextOutgoingPacketResult result = new GameServerGetNextOutgoingPacketResult();

            result.Result = GetNextOutgoingPacket(data, out result.NetAdr, out result.Port);

            return result;


        }

        public void EnableHeartbeats(bool active)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_EnableHeartbeats(active);
        }

        public void SetHeartbeatInterval(int heartbeatInterval)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_SetHeartbeatInterval(heartbeatInterval);
        }

        public void ForceHeartbeat()
        {
            //CheckIfUsable();
            NativeMethods.GameServer_ForceHeartbeat();
        }

        public void AssociateWithClan(SteamID steamIDClan)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_AssociateWithClan(steamIDClan.AsUInt64);
        }

        public void ComputeNewPlayerCompatibility(SteamID steamIDNewPlayer)
        {
            //CheckIfUsable();
            NativeMethods.GameServer_ComputeNewPlayerCompatibility(steamIDNewPlayer.AsUInt64);

        }

    }
}